File: src/main/scala/com/vc/medium/L230.scala
package com.vc.medium


class TreeNode230(var _value: Int) {
  var value: Int = _value
  var left: TreeNode230 = null
  var right: TreeNode230 = null
}

object L230 {
  def kthSmallest(root: TreeNode230, k: Int): Int = {
    var count = k
    var res = 0
    def solve(root: TreeNode230): Unit = {
      if(root.left != null) solve(root.left)
      count -= 1
      if(count == 0) {
        res = root.value
        return
      }
      if(root.right != null) solve(root.right)
    }
    solve(root)
    res
  }
}

File: src/main/java/com/vc/medium/L230.java
package com.vc.medium;

import com.vc.hard.TreeNode;

import java.util.*;

class L230 {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> st = new Stack<>();
        TreeNode current = root;
        while(!st.isEmpty() || current != null) {
            if(current != null) {
                st.push(current);
                current = current.left;
            }
            else {
                current = st.pop();
                if(--k == 0) return current.val;
                current = current.right;
            }
        }
        return -1;
    }
}
